### ISSUE TITLE
Seems to be broken on Raspberry Pi 2

### ISSUE THREAD

**Issue by otbe** [clebert/r-pi-usonic#8]
I would like to use this module on my brand new Raspberry Pi 2, but it seems to be broken (always returns -1).

Im using the same code & pins as on Pi B+.

Can someone confirm my problems?

**Comment by pfnegrini** [comment]
I have the same problem. I checked the hardware settings by running a python script to measure distances and it returned the expected results

**Comment by Klemen1337** [comment]
I don't even get anything...
On install and build I get: ```child_process: customFds option is deprecated, use stdio instead.```

What can I do?

**Comment by charlieclark** [comment]
I'm also having trouble getting this to work on my PI 2. Any updates on this?

**Comment by YesPapa** [comment]
Same thing with the Raspberry PI 2. Is there any other solution with another npm?

**Comment by YesPapa** [comment]
How can we do to make compatible this library with the PI 2?

**Comment by skoold2003** [comment]
I can confirm this issue on the PI 2. I have both a V1 and V2 model next to each other and swapping back and forth works for V1 but not V2. If it's any help, I ran into an issue with another library from V1 to V2 where the dev location on the new PI is in a different path. The git commit page is here https://github.com/quick2wire/quick2wire-gpio-admin/pull/6/files. Hope it helps..

**Comment by incmve** [comment]
It's not the pi2 but the kernel, from 3.18 they changed it.
you could try to disable device-tree
in /boot/config.txt
```
device_tree=
```

**Comment by YesPapa** [comment]
You can find a solution here : https://www.raspberrypi.org/forums/viewtopic.php?f=34&t=112283&p=771200

**Comment by jcodling** [comment]
In short, the memory address space has changed since the Pi2 has a different SOC package. The new memory offset is:

```const off_t offset = 0x3f200000;```

This can be found on line 46 of https://github.com/clebert/r-pi-gpio/blob/master/src/gpio.cc

Change it and ```npm rebuild``` and you'll be good to go.

(This info is also in the forum post linked above by YesPapa but I posted this so that it is all in one spot and easier to find)

**Comment by born2net** [comment]
I can confirm this is not working with pi 2, any chance we cab get detailed instructions on how to fix or to have npm updated?

tx

Sean

**Comment by born2net** [comment]
well the good news I made the offset c changes and it works ... so tx
the bad news is that it crashes after a single iteration and will not work afterwards unless I reboot the device OR re-run the script on some other bogus ports (I guess it releases the exports) and re-run it again on proper ports...

like this:


```
PI-~/javascript->./ultrasonicPing.js
Distance: 3.81 cm
PI-~/javascript->./ultrasonicPing.js
Distance: -1.00 cm
```

any help is appreciated (I hate to fork external python from my node.js if I don't have to...)

**Comment by born2net** [comment]
well I ended up hacking it... kind of ugly but works

```
#!/usr/local/bin/node

var usonic = require('r-pi-usonic');
var sensor = usonic.createSensor(23,18, 650)
setInterval(function(){
   try {
      console.log('Distance: ' + sensor().toFixed(2) + ' cm');
   } catch(e) {}
},1000);
```

any fixes would be appreciated

Sean

**Comment by clebert** [comment]
I don't have access to an Raspberry Pi 2 at the moment. I will provide a fix soon, but i can't test it.

**Comment by clebert** [comment]
Hardware (PI): BCM2708 -> peripherals base address: 0x20000000
Hardware (PI2): BCM2709 -> peripherals base address: 0x3F000000

I must find a way to detect on which hardware the software is running. Then i can set the right memory offset. Perhaps  with `cat /proc/cpuinfo | grep Hardware`

**Comment by jcodling** [comment]
Once I changed the base offset to 0x3f200000 mine works just fine on pin 23-24 on a Pi 2 running node-OPCUA which exposes the ultrasonic sensor to OPC UA clients.

i don't now how future proof it is but I thought of using ```lscpu | grep 'CPU(s):' | rev | cut -d':' -f 1``` to grab the number of cpus since the Pi2 has 4 vs the single core of the previous models.

**Comment by clebert** [comment]
I have updated the master branch. Can someone confirm that my fix works? After confirmation i will publish a new release on NPM. Thanks.

**Comment by Soulskater** [comment]
I've tested on the master branch and it worked! I'm looking forward to the new node package!
Thanks,
Gabor

**Comment by clebert** [comment]
Thanks for testing :+1:  I've just published v2.5.0 on npm.
